server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1032,6 +1038,35 @@ private boolean mayHaveBeenIndexedBefore(Index index) {
         return mayHaveBeenIndexBefore;
     }
 
+    /**
+     * Checks if the document of an append-only request may have been exposed or seen before on replica.
+     * It's ok to return `true` even if this is not the case but returning `false` must always be correct.
+     */
+    private boolean mayHaveBeenExposedBefore(Index index) {
+        /*
+         * As soon as an append-only request was indexed into the primary, it can be exposed to a search then users can issue
+         * a follow up operation on it. In rare cases, the follow up operation can be arrived and processed on a replica before
+         * the original append-only. In this case we can't simply proceed with the append only without consulting the version map.
+         * If a replica has seen a non-append-only operation with a higher seqno than the seqno of an append-only, it may have seen
+         * the document of that append-only request. However, if the seqno of an append-only is higher than seqno of any non-append-only
+         * requests, we can assert the replica have not seen the document of that append-only request, thus we can apply optimization.
+         */
+        assert canOptimizeAddDocument(index);
+        assert index.origin() != Operation.Origin.PRIMARY;
+        return index.seqNo() <= maxSeqNoOfNonAppendOnlyOperations.get();
+    }
+
+    private void updateMaxSeqNoOfNonAppendOnlyOperations(Operation op) {
```

Comment:
maybe just inline this into the `planIndexingAsNonPrimary` method? I think that would be cleaner.